src/wallet/bdb.h

```diff
@@ -207,6 +207,12 @@ class BerkeleyBatch
         operator Dbt*();
     };
 
+private:
+    bool ReadKey(CDataStream& key, CDataStream& value);
+    bool WriteKey(CDataStream& key, CDataStream& value, bool overwrite=true);
+    bool EraseKey(CDataStream& key);
+    bool HasKey(CDataStream& key);
```

Comment:
I found it really confusing that all the `const&` are replaced by `&`. Is there any reason for that? If not, it could make sense to fix this up in a follow-up commit.

Reply:
I agree. If anything, `const` is good documentation about what can potentially be modified and what not.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/19292#discussion_r442219249

> I found it really confusing that all the `const&` are replaced by `&`. Is there any reason for that? If not, it could make sense to fix this up in a follow-up commit.

I assumed it was necessary to remove const from all the arguments because reading from CDataStream objects changes their internal real read position. This is a reason I suggested replacing all the CDataStream arguments with Spans https://github.com/bitcoin/bitcoin/pull/18971#discussion_r437342744. This could be done as a later followup.

Reply:
Both `data` and `size` are marked `const` in the `CDataStream` class. How would this affect the read position anyway?

I think the reason is that the in-memory key data needs to be purged (in case it is a private key)?

Reply:
> Both `data` and `size` are marked `const` in the `CDataStream` class. How would this affect the read position anyway?

Guess it wouldn't. I didn't check if these were const because I just remembered CDataStream being goofy about reading. Maybe it's less goofy than I thought but I still think Spans are better.

Reply:
I agree that spans are better when all that is needed is to pass down `data` and `size`. I'll leave this diff here up for grabs:

```diff
diff --git a/src/wallet/bdb.cpp b/src/wallet/bdb.cpp
index 125bf004e4..46d12546c7 100644
--- a/src/wallet/bdb.cpp
+++ b/src/wallet/bdb.cpp
@@ -804,7 +804,7 @@ std::string BerkeleyDatabaseVersion()
     return DbEnv::version(nullptr, nullptr, nullptr);
 }
 
-bool BerkeleyBatch::ReadKey(CDataStream& key, CDataStream& value)
+bool BerkeleyBatch::ReadKey(Span<char> key, CDataStream& value)
 {
     if (!pdb)
         return false;
@@ -822,7 +822,7 @@ bool BerkeleyBatch::ReadKey(CDataStream& key, CDataStream& value)
     return false;
 }
 
-bool BerkeleyBatch::WriteKey(CDataStream& key, CDataStream& value, bool overwrite)
+bool BerkeleyBatch::WriteKey(Span<char> key, Span<char> value, bool overwrite)
 {
     if (!pdb)
         return true;
@@ -840,7 +840,7 @@ bool BerkeleyBatch::WriteKey(CDataStream& key, CDataStream& value, bool overwrit
     return (ret == 0);
 }
 
-bool BerkeleyBatch::EraseKey(CDataStream& key)
+bool BerkeleyBatch::EraseKey(Span<char> key)
 {
     if (!pdb)
         return false;
@@ -855,7 +855,7 @@ bool BerkeleyBatch::EraseKey(CDataStream& key)
     return (ret == 0 || ret == DB_NOTFOUND);
 }
 
-bool BerkeleyBatch::HasKey(CDataStream& key)
+bool BerkeleyBatch::HasKey(Span<char> key)
 {
     if (!pdb)
         return false;
diff --git a/src/wallet/bdb.h b/src/wallet/bdb.h
index 1e717b95be..f5fa2da76a 100644
--- a/src/wallet/bdb.h
+++ b/src/wallet/bdb.h
@@ -208,10 +208,10 @@ class BerkeleyBatch
     };
 
 private:
-    bool ReadKey(CDataStream& key, CDataStream& value);
-    bool WriteKey(CDataStream& key, CDataStream& value, bool overwrite=true);
-    bool EraseKey(CDataStream& key);
-    bool HasKey(CDataStream& key);
+    bool ReadKey(Span<char> key, CDataStream& value);
+    bool WriteKey(Span<char> key, Span<char> value, bool overwrite = true);
+    bool EraseKey(Span<char> key);
+    bool HasKey(Span<char> key);
 
 protected:
     Db* pdb;

Reply:
@achow101 Let us know if you want to change this to span or rather have the current version of the pr merged

Reply:
It looks like my patch only compiles after commit b8740d6737b49229101fbce909ca3917929185b4 (Merge #18468: Span improvements)

Reply:
I'll leave it for a followup.

Reply:
+1 on using spans
I also thought about making `BerkeleyBatch::ReadKey` return an `Option<CDataStream>` instead of passing it in, but dunno.

Reply:
Done in #19320